File: auto-suggest-ios/AutoSuggestApp/ResponseDetailViewController.m
/*
 * Copyright (c) 2011-2022 HERE Europe B.V.
 * All rights reserved.
 */

#import "ResponseDetailViewController.h"

@interface ResponseDetailViewController () <UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UILabel *labelDescription;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ResponseDetailViewController {
    NSArray<NMALink *>* _discoveryResults;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    // perform detail request for these types of autoSuggest
    if ([self.autoSuggest isKindOfClass:[NMAAutoSuggestPlace class]]) {
        NMAAutoSuggestPlace* place = (NMAAutoSuggestPlace *) self.autoSuggest;
        [place.placeDetailsRequest startWithBlock:^(NMARequest* request, id data, NSError* error) {
            if (data) {
                NMAPlace* place = (NMAPlace*)data;
                NSMutableString* lableText = [NSMutableString new];
                [lableText appendFormat:@"Name: %@",  place.name];
                [lableText appendFormat:@"\nAlternative names:"];

                for (NSString* key in place.alternativeNames) {
                    [lableText appendFormat:@"\n%@=%@", key, [place.alternativeNames objectForKey:key]];
                }
                self.labelDescription.text = lableText;
            } else {
                NSLog(@"placeDetailsRequest returns error with error code:%d", (int)error.code);
            }
        }];
    } else if ([self.autoSuggest isKindOfClass:[NMAAutoSuggestSearch class]]) {
        self.tableView.hidden = NO;
        self->_tableView.dataSource = self;
        NMAAutoSuggestSearch* search = (NMAAutoSuggestSearch *) self.autoSuggest;
        [search.suggestedSearchRequest startWithBlock:^(NMARequest* request, id data, NSError* error) {
            if (data) {
                NMADiscoveryPage* page = (NMADiscoveryPage*)data;
                self->_discoveryResults = page.discoveryResults;
                [self->_tableView reloadData];
            } else {
                NSLog(@"suggestedSearchRequest returns error with error code:%d", (int)error.code);
            }
        }];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _discoveryResults.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
    }

    cell.textLabel.numberOfLines = 1;
    cell.detailTextLabel.numberOfLines = 5;
    [cell.textLabel sizeToFit];

    NMALink* item = [_discoveryResults objectAtIndex:indexPath.row];
    cell.textLabel.text = [NSString stringWithFormat:@"Name:%@", item.name];

    if ([item isKindOfClass:NMAPlaceLink.class]) {
        NSMutableString* text = [NSMutableString new];
        NMAPlaceLink* placeLink = (NMAPlaceLink*)item;
        [text appendFormat:@"%@Vicinity:%@", text, placeLink.vicinityDescription];
        [text appendFormat:@"%@\nCategory:%@", text, placeLink.category.name];
        [text appendFormat:@"%@\nDistance:%ld", text, placeLink.distance];
        cell.detailTextLabel.text = text;
    }

    return cell;
}

@end

File: auto-suggest-ios-swift/AutoSuggestApp/ResponseDetailViewController.swift
/*
 * Copyright (c) 2011-2022 HERE Europe B.V.
 * All rights reserved.
 */

import UIKit
import NMAKit

class ResponseDetailViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {

    var autoSuggests:NMAAutoSuggest!

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var responseDescription: UILabel!

    var searchResultData = [NMALink]()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // perform detail request for these types of autoSuggest
        if let place = autoSuggests as? NMAAutoSuggestPlace {
            place.placeDetailsRequest().start({ request, data, inError in

                guard inError == nil else {
                    print("placeDetailsRequest returns error with error code:\((inError! as NSError).code)")
                    return
                }

                let response:NMAPlace = data as! NMAPlace
                var lableText = String(format:"name: %@", response.name ?? "")
                lableText = String(format:"%@\nAlternative names:", lableText)

                for (key, val) in response.alternativeNames ?? [:] {
                    lableText = String(format:"%@\n%@=%@", lableText, key, val)
                }

                self.responseDescription.text = lableText

            })
        } else if let searchRequest = autoSuggests as? NMAAutoSuggestSearch {
            tableView.isHidden = false
            tableView.dataSource = self

            searchRequest.suggestedSearchRequest().start({ request, data, inError in
                guard inError == nil else {
                    print("suggestedSearchRequest returns error with error code:\((inError! as NSError).code)")
                    return
                }
                
                if let page = data as? NMADiscoveryPage {
                    self.searchResultData = page.discoveryResults
                    self.tableView.reloadData()
                }
            })
        }
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return searchResultData.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier:"cell")
        if cell == nil {
            cell = UITableViewCell(style:.subtitle, reuseIdentifier:"cell")
        }

        cell?.textLabel?.numberOfLines = 1
        cell?.detailTextLabel?.numberOfLines = 5
        cell?.textLabel?.sizeToFit()

        let item = searchResultData[indexPath.row]
        cell?.textLabel?.text = String(format:"Name:%@", item.name ?? "")

        if let placeLink = item as? NMAPlaceLink {
            var text = String(format:"Vicinity:%@", placeLink.vicinityDescription ?? "")
            text = String(format:"%@\nCategory:%@", text, placeLink.category?.name ?? "")
            text = String(format:"%@\nDistance:%ld", text, placeLink.distance)
            cell?.detailTextLabel?.text = text
        }

        return cell!
    }
}
